www/_blog/2022-08-11-authentication-tutorial.mdx

```diff
@@ -0,0 +1,597 @@
+
+```html
+<!-- in "templates/index.html" -->
+{% extends "base.html" %}
+{% block content %}
+<h1>Hello world</h1>
+{% if username %}
+    <h3>Logged in: {{ username }}</h3>
+{% endif %}
+{% endblock content %}
+```
+
+Using our middleware in requests is easy in Axum by including a reference to it in the parameters. We then add the username to the context for it to be rendered on the page.
+
+```rust
+async fn index(
+    Extension(current_user): Extension<AuthState>,
+    Extension(templates): Extension<Templates>,
+) -> impl IntoResponse {
+    let mut context = Context::new();
+    if let Some(user) = current_user.get_user().await {
+        context.insert("username", &user.username);
+    }
+    Html(templates.render("index", &context).unwrap())
+}
+```
+
+### Logging in and logging out
+
+Great we can signup and that now puts us in a session. We may want to log out and drop the session. This is very simple to do by returning a response with the cookie `Max-Age` set to 0.
+
+```rust
+pub(crate) async fn logout_response() -> impl axum::response::IntoResponse {
+    Response::builder()
+        .status(http::StatusCode::SEE_OTHER)
+        .header("Location", "/")
+        .header("Set-Cookie", "session_token=_; Max-Age=0")
+        .body(Empty::new())
+        .unwrap()
+}
+```
+
+For logging in we have a very similar logic for signup with pulling multipart information of a post request. Unlike signup, we don't want to create a new user. We want to check there is an existing user with a matching password and then create a new session:
+
+```rust
+async fn post_login(
+    Extension(database): Extension<Database>,
+    multipart: Multipart,
+) -> impl IntoResponse {
+    let data = parse_multipart(multipart)
+        .await
+        .map_err(|err| error_page(&err))?;
+
+    if let (Some(username), Some(password)) = (data.get("username"), data.get("password")) {
+        const LOGIN_QUERY: &str = "SELECT id, password FROM users WHERE users.username = $1;";
+
+        let row: Option<(i32, String)> = sqlx::query_as(LOGIN_QUERY)
+            .bind(username)
+            .fetch_optional(database)
+            .await
+            .unwrap();
+
+        let (user_id, hashed_password) = if let Some(row) = row {
+            row
+        } else {
+            return Err(LoginError::UserDoesNotExist);
+        };
+
+        // Verify password against PHC string
+        let parsed_hash = PasswordHash::new(&hashed_password).unwrap();
+        if let Err(_err) = Pbkdf2.verify_password(password.as_bytes(), &parsed_hash) {
+            return Err(LoginError::WrongPassword);
+        }
+
+        let session_token = new_session(database, random, user_id);
+
+        let resp = http::Response::builder()
+            .status(http::StatusCode::SEE_OTHER)
+            .header("Location", "/")
+            .header("Set-Cookie", format!("session_token={}; Max-Age=999999", session_token))
+            .body(http_body::Empty::new())
+            .unwrap()

```

Comment:
Maybe there should be a function (`set_cookie`) with this code for both handlers to use?